### DISCUSSION TITLE
FEATURE ANNOUNCEMENT: zstd compression for HTTP

### DISCUSSION THREAD

**Discussion by ppekrol** [ravendb/ravendb#17920] [Announcements]
Affected versions:
6.0.2+

Affected subsystems:
HTTP API

Affected clients:
C# Client API

As a continuation of our work with [zstd compression algorithm](https://github.com/facebook/zstd) we were encouraged from previous [Backup & Export results](https://github.com/ravendb/ravendb/discussions/17678) and went deeper to the HTTP API itself. This post is sponsored by our findings and summarizes the new feature.

## Bulk Insert

First let's look at bulk insert performance in couple of scenarios and using not only gzip and zstd, but also [brotli algorithm](https://github.com/google/brotli). All of them with various compression levels.

| Algorithm | Compression Level | Items | Time (s) | Items/s | Items/s (%) | Written (bytes) | Written (MB) |
| ------- | ------------------ | ------- | ------------------ | ----- | ----- | ----- | ----- |
| Gzip   | Fastest | 100_000 | 1.29 | 77,366.33 | 0.00% | 588,636 | 0.56 | 
| Gzip   | Optimal | 100_000 | 1.39 | 71,897.84 | -7.61% | 568,750 | 0.54 |
| Brotli   | Fastest | 100_000 | 1.93 | 51,771.89 | -49.44% | 25,302,139 | 24.13 | 
| Brotli   | Optimal | 100_000 | 1.49 | 67,318.69 | -14.93% | 229,867 | 0.22 | 
| Zstd   | Fastest | 100_000 | 1.27 | 78,975.14 | 2.04% | 259,563 | 0.25 | 
| NoCompression   | N/A | 100_000 | 1.39 | 71,767.04 | -7.80% | 35,179,725 | 33.55 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 1_000_000 | 11.19 | 89,398.43 | 0.00% | 5,882,511 | 5.61 | 
| Gzip   | Optimal | 1_000_000 | 11.64 | 85,888.74 | -4.09% | 5,683,282 | 5.42 | 
| Brotli   | Fastest | 1_000_000 | 18.29 | 54,683.86 | -63.48% | 254,552,310 | 242.76 | 
| Brotli   | Optimal | 1_000_000 | 12.43 | 80,465.01 | -11.10% | 2,327,839 | 2.22 | 
| Zstd   | Fastest | 1_000_000 | 11.07 | 90,304.61 | 1.00% | 2,285,896 | 2.18 |
| NoCompression   | N/A | 1_000_000 | 11.71 | 85,421.02 | -4.66% | 353,779,057 | 337.39 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 10_000_000 | 109.53 | 91,301.37 | 0.00% | 58,856,571 | 56.13 | 
| Gzip   | Optimal | 10_000_000 | 117.72 | 84,946.76 | -4.09% | 56,864,276 | 54.23 | 
| Brotli   | Fastest | 10_000_000 | 180.40 | 55,431.73 | -63.48% | 2,545,840,000 | 2,427.90 | 
| Brotli   | Optimal | 10_000_000 | 121.05 | 82,611.02 | -11.10% | 23,236,444 | 22.16 | 
| Zstd   | Fastest | 10_000_000 | 108.12 | 92,494.02 | 1.00% | 22,691,185 | 21.64 | 
| NoCompression   | N/A | 10_000_000 | 130.66 | 76,534.94 | -4.66% | 3,557,310,000 | 3,392.52 | 

![Docs per second](https://github.com/ravendb/ravendb/assets/446555/14942f95-2b47-4da7-9ddf-689fa28b5c97)
![Data Written (in MB)](https://github.com/ravendb/ravendb/assets/446555/c4acfc6c-efce-4844-8103-d4168f125577)

Couple of observations from this scenario:
- NoCompression, despite running locally, does not yield best results and the amount of data transfered is tremendous
- Gzip (Fastest) is quite good when it comes to performance and Gzip (Optimal) is a bit better when it comes to amount of data sent, but probably not worth the performance drop
- Brotli (Fastest) has some serious issues for both performance and amount of data sent, should be banned by law, we could not figure out why this is happening
- Both Brotli (Optimal) and Zstd (Fastest) are excellent when it comes to compression and are yielding 55-60% less data over the wire with Zstd (Fastest) compressing much faster

## Session (small)

No let's jump into Session with small documents (single property), we are testing similar scenario to bulk insert with single batch and different batch sizes. Since Brotli (Fastest) was performing poorly, we removed it from the equation entirely.

| Algorithm | Compression Level | Items in Batch | Time (s) | Items/s | Items/s (%) | Written (bytes) | Written (MB) |
| ------- | ------------------ | ------- | ------------------ | ----- | ----- | ----- | ----- |
| Gzip   | Fastest | 1 | 0.02 | 47.84 | 0.00% | 268 | 0.00 | 
| Brotli   | Optimal | 1 | 0.02 | 47.72 | -0.25% | 241 | 0.00 | 
| Zstd   | Fastest | 1 | 0.02 | 52.09 | 8.16% | 264 | 0.00 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 10 | 0.02 | 427.88 | 0.00% | 338 | 0.00 | 
| Brotli   | Optimal | 10 | 0.02 | 458.80 | 6.74% | 293 | 0.00 | 
| Zstd   | Fastest | 10 | 0.02 | 456.64 | 6.30% | 339 | 0.00 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 100 | 0.04 | 2,272.25 | 0.00% | 906 | 0.00 | 
| Brotli   | Optimal | 100 | 0.04 | 2,324.28 | 2.24% | 598 | 0.00 | 
| Zstd   | Fastest | 100 | 0.05 | 2,193.29 | -3.60% | 658 | 0.00 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 1_000 | 0.22 | 4,543.12 | 0.00% | 6,154 | 0.01 | 
| Brotli   | Optimal | 1_000 | 0.23 | 4,443.44 | -2.24% | 2,938| 0.00 | 
| Zstd   | Fastest | 1_000 | 0.22 | 4,466.26 | -1.72% | 3,399 | 0.00 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 10_000 | 0.68 | 14,760.05 | 0.00% | 59,125 | 0.06 | 
| Brotli   | Optimal | 10_000 | 0.71 | 14,152.07 | -4.30% | 26,531 | 0.03 | 
| Zstd   | Fastest | 10_000 | 0.61 | 16,307.04 | 9.49% | 30,167 | 0.03 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 100_000 | 5.10 | 19,592.33 | 0.00% | 589,547 | 0.56 | 
| Brotli   | Optimal | 100_000 | 5.01 | 19,971.95 | 1.90% | 256,747 | 0.24 | 
| Zstd   | Fastest | 100_000 | 5.26 | 18,997.71 | -3.13% | 235,530 | 0.22 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 1_000_000 | 63.44 | 15,762.89 | 0.00% | 5,892,997.00 | 5.62 | 
| Brotli   | Optimal | 1_000_000 | 61.33 | 16,305.13 | 3.33% | 2,600,468.00 | 2.48 | 
| Zstd   | Fastest | 1_000_000 | 61.81 | 16,179.58 | 2.58% | 2,548,040.00 | 2.43 | 
|    |  |  |  |  |  |  |  |  |

We think it it safe to say that performance here is within a margin of an error and all 3 algorithms are behaving more than fine, but when it comes to the data sent, both Brotli (Optimal) and Zstd (Fastest) are transferring much less data over the wire (up to 60%). This should be major factor when considering latency, even in fast network environments.

## Session (medium)

Last scenario is a mirror of a second one, but we are creating a more complex object, with an array of items (filling out Order from the sample Northwind database).

| Algorithm | Compression Level | Items in Batch | Time (s) | Items/s | Items/s (%) | Written (bytes) | Written (MB) |
| ------- | ------------------ | ------- | ------------------ | ----- | ----- | ----- | ----- |
| Gzip   | Fastest | 1 | 0.14 | 7.10 | 0.00% | 320 | 0.00 | 
| Brotli   | Optimal | 1 | 0.14 | 7.22 | 1.66% | 304 | 0.00 | 
| Zstd   | Fastest | 1 | 0.14 | 7.07 | -0.47% | 330 | 0.00 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 10 | 0.14 | 69.29 | 0.00% | 597 | 0.00 | 
| Brotli   | Optimal | 10 | 0.14 | 69.70 | 0.58% | 496 | 0.00 | 
| Zstd   | Fastest | 10 | 0.15 | 68.79 | -0.73% | 522 | 0.00 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 100 | 0.18 | 557.01 | 0.00% | 3,123 | 0.00 | 
| Brotli   | Optimal | 100 | 0.18 | 557.27 | 0.05% | 2,099 | 0.00 | 
| Zstd   | Fastest | 100 | 0.18 | 555.09 | -0.35% | 2,181 | 0.00 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 1_000 | 0.42 | 2,383.95 | 0.00% | 27,514.88 | 0.03 | 
| Brotli   | Optimal | 1_000 | 0.42 | 2,387.67 | 0.16% | 18,575.36 | 0.02 | 
| Zstd   | Fastest | 1_000 | 0.40 | 2,484.88 | 4.06% | 19,486.72 | 0.02 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 10_000 | 2.09 | 4,785.33 | 0.00% | 272,721.92 | 0.26 | 
| Brotli   | Optimal | 10_000 | 2.27 | 4,410.79 | -8.49% | 191,447.04 | 0.18 | 
| Zstd   | Fastest | 10_000 | 1.75 | 5,709.84 | 16.19% | 208,762.88 | 0.20 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 100_000 | 7.88 | 12,697.00 | 0.00% | 2,715,812.00 | 2.59 | 
| Brotli   | Optimal | 100_000 | 7.91 | 12,647.04 | -0.40% | 1,929,380.00 | 1.84 | 
| Zstd   | Fastest | 100_000 | 7.93 | 12,615.64 | -0.64% | 2,107,638.00 | 2.01 | 
|    |  |  |  |  |  |  |  |  |
| Gzip   | Fastest | 1_000_000 | 80.86 | 12,366.98 | 0.00% | 26,770,145.00 | 25.53 | 
| Brotli   | Optimal | 1_000_000 | 71.31 | 14,023.84 | 11.81% | 19,294,000.00 | 18.40 | 
| Zstd   | Fastest | 1_000_000 | 69.34 | 14,420.90 | 14.24% | 20,583,547.00 | 19.63 | 
|    |  |  |  |  |  |  |  |  |

Similarly to previous scenario, the amount of data sent is considerably lower both for Brotli (Optimal) and Zstd (Fastest). 

## Conclusion

If we will look at both types of scenarios, bulk insert and session, we can make a conclusion that beside some hiccups (yes, you Brotli Fastest), both Brotli and Zstd have better compression ratios than Gzip and should be considered as a replacement for Gzip.  

At the end, to lower the code complexity and due to better performance during Bulk Insert operations, we decided to pick Zstd as the future compression algorithm of our choice, same as AWS did (more [here](https://www.infoq.com/news/2022/09/amazon-gzip-zstd/)) and we happy to say that our internal APIs are using Zstd for over a month now and we see significant (~25%) decrease in amount of data sent over the wire.

But for now, we are leaving Gzip as the default with the ability to switch to Zstd by using `DocumentConventions.HttpCompressionAlgorithm' property. And we will consider switching to Zstd as default (with the ability to switch back to Gzip of course) in next major version of our product.

We encourage each one of you to take Zstd for a spin, but remember to update both Server and Client API first :)

Enjoy!
RavenDB Team

**Comment by kamranayub** [comment]
This is awesome news, thanks for this! 👏 I'll try to measure the difference in my app and share that too.
